src/planner/partialize.c

```diff
@@ -185,6 +197,656 @@ partialize_agg_paths(RelOptInfo *rel)
+	if (parse->groupingSets)
+		return;
+
+	/* Don't replan aggregation if we already have a MinMaxAggPath (e.g., created by
+	 * ts_preprocess_first_last_aggregates) */
+	if (has_min_max_agg_path(output_rel))
+		return;
+
+	/* Is sorting possible ? */
+	bool can_sort = grouping_is_sortable(parse->groupClause) &&
+					ts_guc_enable_partitionwise_plain_sorted_aggregation;
+
+	/* Is hashing possible ? */
+	bool can_hash = grouping_is_hashable(parse->groupClause) &&
+					!ts_is_gapfill_path(linitial(output_rel->pathlist)) &&
+					ts_guc_enable_partitionwise_hashed_aggregation;
+
+	Assert(extra != NULL);
+	GroupPathExtraData *extra_data = (GroupPathExtraData *) extra;
+
+	/* Determine the number of groups from the already planned aggregation */
+	AggPath *existing_agg_path = get_existing_agg_path(output_rel);
+	if (existing_agg_path == NULL)
+		return;
+
+	/* Skip partial aggregations already created by _timescaledb_functions.partialize_agg */
+	if (existing_agg_path->aggsplit == AGGSPLIT_INITIAL_SERIAL)
+		return;
+
+/* Don't replan aggregation if it contains already partials or non-serializable aggregates */
+#if PG14_LT
+	AggClauseCosts agg_costs;
+	MemSet(&agg_costs, 0, sizeof(AggClauseCosts));
+	get_agg_clause_costs_compat(root, (Node *) root->processed_tlist, AGGSPLIT_SIMPLE, &agg_costs);
+	get_agg_clause_costs_compat(root, parse->havingQual, AGGSPLIT_SIMPLE, &agg_costs);
+
+	if (agg_costs.hasNonPartial || agg_costs.hasNonSerial)
+#else
+	if (root->hasNonPartialAggs || root->hasNonSerialAggs)
+#endif
+		return;
+
+	double d_num_groups = existing_agg_path->numGroups;
+	Assert(d_num_groups > 0);
+
+	/* Construct partial group agg upper relation */
+	RelOptInfo *partially_grouped_rel =
+		fetch_upper_rel(root, UPPERREL_PARTIAL_GROUP_AGG, input_rel->relids);
+	partially_grouped_rel->consider_parallel = input_rel->consider_parallel;
+	partially_grouped_rel->reloptkind = input_rel->reloptkind;
+	partially_grouped_rel->serverid = input_rel->serverid;
+	partially_grouped_rel->userid = input_rel->userid;
+	partially_grouped_rel->useridiscurrent = input_rel->useridiscurrent;
+	partially_grouped_rel->fdwroutine = input_rel->fdwroutine;
+
+	/* Build target list for partial aggregate paths */
+	PathTarget *grouping_target = output_rel->reltarget;
+	PathTarget *partial_grouping_target = ts_make_partial_grouping_target(root, grouping_target);
+	partially_grouped_rel->reltarget = partial_grouping_target;
+
+	/* Calculate aggregation costs */
+	if (!extra_data->partial_costs_set)
+	{
+		/* partial phase */
+		get_agg_clause_costs_compat(root,
+									(Node *) partial_grouping_target->exprs,
+									AGGSPLIT_INITIAL_SERIAL,
+									&extra_data->agg_partial_costs);
+
+#if PG14_LT
+		PathTarget *target = root->upper_targets[UPPERREL_GROUP_AGG];
+#endif
+
+		/* final phase */
+		get_agg_clause_costs_compat(root,
+									(Node *) target->exprs,
+									AGGSPLIT_FINAL_DESERIAL,
+									&extra_data->agg_final_costs);
```

Comment:
`target` is not defined on PG >= 14, but is still used somehow in `get_agg_clause_costs_compat`? This looks somewhat confusing, maybe we should pass `root->upper_targets[UPPERREL_GROUP_AGG]` as an argument directly, it's available in all versions anyway.